public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts

```diff
@@ -0,0 +1,151 @@
+import { EntryType, Row, RowGroup } from '../components/ResourcePicker/types';
+import { SUPPORTED_LOCATIONS, SUPPORTED_RESOURCE_TYPES } from './supportedResources';
+
+const RESOURCE_GRAPH_URL =
+  '/resourcepickerdata/providers/Microsoft.ResourceGraph/resources?api-version=2020-04-01-preview';
+
+type RawAzureResourceGraphDataItem = {
+  type: string;
+  subscriptionId: string;
+  name: string;
+  id: string;
+  resourceGroup: string;
+  location: string;
+};
+
+export default class ResourcePickerData {
+  private proxyUrl: string;
+  private formattedResourcePickerData: RowGroup;
+
+  constructor(proxyUrl?: string) {
+    this.proxyUrl = proxyUrl!;
+    this.formattedResourcePickerData = {};
+  }
+
+  async getResourcePickerData() {
+    // returns whatever is cached in memory, only loads data if it has none
+    const hasPreviouslyLoadedResourcePickerData = Object.keys(this.formattedResourcePickerData).length > 0;
+    if (hasPreviouslyLoadedResourcePickerData) {
+      return this.formattedResourcePickerData;
+    }
+
+    const { ok, data: response } = await this._makeResourceGraphRequest('resourcecontainers');
+
+    // TODO: figure out desired error handling strategy
+    if (!ok) {
+      throw new Error('unable to fetch resource containers');
+    }
+
+    return this._formatAndCacheResourceData(response.data);
+  }
+
+  async getResourcePickerDataWithNestedResourceData(resourceGroup: Row) {
+    // subscription data should already be pre-loaded
+    if (resourceGroup.typeLabel === 'Subscription') {
+      return this.formattedResourcePickerData;
+    }
+
+    // check to see if we've already loaded nested resources for this resource group
+    // if we have, return whatever is in memory
+    const subscriptionObj = this.formattedResourcePickerData[resourceGroup.subscriptionId];
+    const resourceGroupObj = subscriptionObj?.children && subscriptionObj.children[resourceGroup.name.toLowerCase()];
+    const hasPreviouslyLoadedResoucesForResourceGroup =
+      resourceGroupObj?.children && Object.keys(resourceGroupObj.children).length > 0;
+    if (hasPreviouslyLoadedResoucesForResourceGroup) {
+      return this.formattedResourcePickerData;
+    }
+
+    const { ok, data: response } = await this._makeResourceGraphRequest(`
+      resources 
+      | where resourceGroup == "${resourceGroup.name.toLowerCase()}"
+      | where type in (${SUPPORTED_RESOURCE_TYPES}) and location in (${SUPPORTED_LOCATIONS})
+    `);
+
+    // TODO: figure out desired error handling strategy
+    if (!ok) {
+      throw new Error('unable to fetch resource containers');
+    }
+
+    return this._formatAndCacheResourceData(response.data);
+  }
+
+  _formatAndCacheResourceData(rawData: RawAzureResourceGraphDataItem[]) {
+    // "produce" from immer takes in an object, allows you to edit a copy of that object and then returns a new object
+    this.formattedResourcePickerData = produce(this.formattedResourcePickerData, (draftState: any) => {
```

Comment:
hmmm... one thing that I've been trying to get away from is storing less state like this in these classes, and store/produce what we need in the react components themselves.

I think maintaining request caches in these classes is fine, but I think it would be simpler if these functions were a bit more... "functional" and just mapped over the data and returned the new format.

the `getResourcePickerDataWithNestedResourceData` function could then just be like `getResources`, returns just the list of resources, and then `fetchNested` back in the react component inserts them as the children. Thoughts?

Reply:
Yeah I guess I liked the idea that we wouldn't make as many calls to the backend even if you switched between logs/metrics but you're probably right that it's weird to have this stuff hanging around hogging up memory even when the component is gone.